driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -196,17 +193,13 @@ private PoolingOptions options() {
         return manager.configuration().getPoolingOptions();
     }
 
-    public Connection borrowConnection(long timeout, TimeUnit unit) throws ConnectionException, TimeoutException {
+    ListenableFuture<Connection> borrowConnection(int maxQueueSize) {
         Phase phase = this.phase.get();
         if (phase != Phase.READY)
-            // Note: throwing a ConnectionException is probably fine in practice as it will trigger the creation of a new host.
-            // That being said, maybe having a specific exception could be cleaner.
-            throw new ConnectionException(host.getSocketAddress(), "Pool is " + phase);
+            return Futures.immediateFailedFuture(new ConnectionException(host.getSocketAddress(), "Pool is " + phase));
 
         if (connections.isEmpty()) {
-            if (!host.convictionPolicy.canReconnectNow())
-                throw new TimeoutException("Connection pool is empty, currently trying to reestablish connections");
-            else {
+            if (host.convictionPolicy.canReconnectNow()) {
                 int coreSize = options().getCoreConnectionsPerHost(hostDistance);
                 if (coreSize == 0) {
                     maybeSpawnNewConnection();
```

Comment:
Well it is part of `borrowConnection()` which if I'm not mistaken is called for every request so even if it does one at a time, the call here only checks on options().getCoreConnectionsPerHost() which does not check the number of currently open connections, so it looks like we would keep on opening connections (one at a time) until the max is reached, is that correct? Not saying it's a bad behaviour, was just seeing that we don't mention that anywhere in docs